zeus/metric.py

```diff
@@ -0,0 +1,532 @@
+
+
+class EnergyCumulativeCounter(Metric):
+    """EnergyCumulativeCounter class to monitor and record cumulative energy consumption.
+
+    This class tracks GPU, CPU, and DRAM energy usage over time, and records the data as Prometheus Counter metrics.
+    The energy consumption metrics are periodically updated and pushed to a Prometheus Push Gateway for monitoring and analysis.
+
+    The cumulative nature of the Counter ensures that energy values are always incremented over time, never reset,
+    which is ideal for tracking continuously increasing values like energy usage.
+
+    Attributes:
+        energy_monitor: The ZeusMonitor instance that collects energy consumption data for the system.
+        update_period: The interval (in seconds) between consecutive energy data updates.
+        prometheus_url: The URL of the Prometheus Push Gateway where the Counter metrics will be pushed.
+        job: The name of the job associated with the energy monitoring in Prometheus.
+        gpu_counters: A dictionary storing the Prometheus Counter metrics for each GPU.
+        cpu_counters: A dictionary storing the Prometheus Counter metrics for each CPU.
+        dram_counters: A dictionary storing the Prometheus Counter metrics for DRAM.
+        queue: A multiprocessing queue used to send signals to start/stop energy monitoring.
+        proc: A multiprocessing process that runs the energy monitoring loop.
+    """
+
+    def __init__(
+        self,
+        cpu_indices: list,
+        gpu_indices: list,
+        update_period: int,
+        prometheus_url: str,
+        job: str,
+    ) -> None:
+        """Initialize the EnergyCumulativeCounter.
+
+        Args:
+            cpu_indices (list): List of CPU indices to monitor.
+            gpu_indices (list): List of GPU indices to monitor.
+            update_period: The time interval (in seconds) at which energy measurements are updated.
+            prometheus_url: The URL for the Prometheus Push Gateway where the metrics will be pushed.
+            job: The name of the job to be associated with the Prometheus metrics.
+        """
+        self.cpu_indices = cpu_indices
+        self.gpu_indices = gpu_indices
+        self.update_period = update_period
+        self.prometheus_url = prometheus_url
+        self.job = job
+        self.gpu_counters = {}
+        self.cpu_counters = {}
+        self.dram_counters = {}
+        self.queue = None
+        self.proc = None
+
+    def begin_window(self, name: str) -> None:
+        """Begin the energy monitoring window.
+
+        Starts a new multiprocessing process that monitors energy usage periodically
+        and pushes the results to the Prometheus Push Gateway.
+
+        Args:
+            name (str): The unique name of the measurement window. Must match between calls to 'begin_window' and 'end_window'.
+        """
+        context = mp.get_context("spawn")
+        self.queue = context.Queue()
+        self.proc = context.Process(
```

Comment:
This will break when users have more than one concurrent monitoring windows, as the second call to `begin_window` will overwrite `self.queue` and `self.proc`. **An arbitrary number of windows can overlap in time in arbitrary ways.** That is why `ZeusMonitor` keeps an internal state for each window with a `dict` key'ed by the name of the window. For this metric, the state of the window is probably the queue and process handles.

This comment also applies to `PowerGauge`.

Reply:
Something like this:

```python
@dataclass
class MonitoringProcessState:
    queue: mp.Queue
    proc: mp.Process

# Inside __init__
self.window_state: dict[str, MonitoringProcessState] = {}

# Inside `begin_window`
if window_name in self.window_state:
    raise ValueError(f"Energy cumulative counter metric '{window_name}' already exists")
queue = ...
proc = ...
self.window_state[window_name] = MonitoringProcessState(queue, proc)
```